### PR TITLE
WaveShaper node implementation without oversampling

### PR DESCRIPTION
**Pull request by khodzha**
Issue at hand: servo/media#205

### CODE REVIEW

**Comment by khodzha** [comment]
im not sure about [this part of the spec](https://webaudio.github.io/web-audio-api/#WaveShaperNode-attributes), specifically whether curve attr can be set exactly once per node lifetime or it can be set/unset multiple times

right now it is implemented so it can be set exactly once and then it `panic!`s on subsequent not-None modifications

File: audio/wave_shaper_node.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+#[derive(Clone, Debug)]
+pub enum WaveShaperNodeMessage {
+    SetCurve(WaveShaperCurve),
+}
+
+#[derive(AudioNodeCommon)]
+pub(crate) struct WaveShaperNode {
+    curve_set: bool,
+    curve: WaveShaperCurve,
+    oversample: OverSampleType,
```
Comment (ferjm) [inline audio/wave_shaper_node.rs]: Let's remove the build time warning and leave a TODO here with a link to a new issue, please.
```rust
#[allow(dead_code)]
// TODO implement tail-time based on the oversample attribute.
// https://github.com/servo/media/issues/XXX
oversample: OverSampleType,
```

File: examples/Cargo.toml [hunk 2]
```diff
@@ -115,3 +115,7 @@ path = "set_value_curve.rs"
 [[bin]]
 name = "stereo_panner"
 path = "stereo_panner.rs"
+
+[[bin]]
+name = "waveshaper"
```
Comment (ferjm) [inline examples/Cargo.toml]: uber-nit: `wave_shaper`

File: audio/wave_shaper_node.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+#[derive(Clone, Debug)]
+pub enum WaveShaperNodeMessage {
+    SetCurve(WaveShaperCurve),
+}
+
+#[derive(AudioNodeCommon)]
+pub(crate) struct WaveShaperNode {
+    curve_set: bool,
```
[the remark this answers is not in the document]
Comment (khodzha) [reply audio/wave_shaper_node.rs]: This is the part i'm not sure about, [see comment](https://github.com/servo/media/pull/276#issuecomment-508918168)

[the remark this answers is not in the document]
Comment (khodzha) [reply audio/wave_shaper_node.rs]: So reading https://github.com/WebAudio/web-audio-api/pull/1653
>  The original intent was that you can only set the buffer(curve) just once, but you can set them to null as many times as you want, but you can't then set it to a non-null buffer.

Thus is_some() is not enough, you can set it to curve, then to None, then try setting it to another curve again and this attempt must fail, am I right?

[the remark this answers is not in the document]
Comment (ferjm) [reply audio/wave_shaper_node.rs]: Ok, fair enough. Thanks for following-up on this.

**Comment by ferjm** [comment]
Thanks!

@bors-servo r+

**Pull request merged**